packages/tokamak/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol

```diff
@@ -54,4 +110,152 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, ISemver {
     function _isUnsafeTarget(address _target) internal view override returns (bool) {
         return _target == address(this) || _target == address(PORTAL);
     }
+
+    /// @notice Sends a deposit native token message to some target address on the other chain. Note that if the call
+    ///         always reverts, then the message will be unrelayable, and any ETH sent will be
+    ///         permanently locked. The same will occur if the target on the other chain is
+    ///         considered unsafe (see the _isUnsafeTarget() function).
+    /// @param _target      Target contract or wallet address.
+    /// @param _amount      Amount of deposit native token.
+    /// @param _message     Message to trigger the target address with.
+    /// @param _minGasLimit Minimum gas limit that the message can be executed with.
+    function sendNativeTokenMessage(
+        address _target,
+        uint256 _amount,
+        bytes calldata _message,
+        uint32 _minGasLimit
+    )
+        external
+    {
+        // Triggers a message to the other messenger. Note that the amount of gas provided to the
+        // message is the amount of gas requested by the user PLUS the base gas value. We want to
+        // guarantee the property that the call to the target contract will always have at least
+        // the minimum gas limit specified by the user.
+        _sendMessage(
+            OTHER_MESSENGER,
+            baseGas(_message, _minGasLimit),
+            _amount,
+            abi.encodeWithSelector(
+                this.relayMessage.selector, messageNonce(), msg.sender, _target, _amount, _minGasLimit, _message
+            )
+        );
+
+        emit SentMessage(_target, msg.sender, _message, messageNonce(), _minGasLimit);
+        emit SentMessageExtension1(msg.sender, _amount);
+
+        unchecked {
+            ++msgNonce;
+        }
+    }
+
+    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only
+    ///         be executed via cross-chain call from the other messenger OR if the message was
+    ///         already received once and is currently being replayed.
+    /// @param _nonce       Nonce of the message being relayed.
+    /// @param _sender      Address of the user who sent the message.
+    /// @param _target      Address that the message is targeted at.
+    /// @param _value       Native token value to send with the message.
+    /// @param _minGasLimit Minimum amount of gas that the message can be executed with.
+    /// @param _message     Message to send to the target.
+    function relayMessage(
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        uint256 _value,
+        uint256 _minGasLimit,
+        bytes calldata _message
+    )
+        external
+        payable
+        override
+    {
+        if (_value > 0) {
```

Comment:
I think we have to assue that user call relayMessage directly. I think it's good to move in `if(_isOtherMessenger)` and change source to `msg.sender` from `PORTAL` in safeTransferFrom.

In fact, according to https://www.notion.so/onther/First-Audit-commit-b1299e2d6d0b31e7626eb06c43fbad5f36b8a1d6-8becbb08eec44fc2888f61097cbbbe2e?pvs=4#4ef5d14b53b547f886687e3ee57d388e, I think this is not problem in case of TON. But if we use another native token. It will make a problem. 🙇🏼‍♂️

Reply:
Yes, I already assumed that this function is called by any addresses. If we move in if(_isOtherMessenger), the function cannot be called successfully by other users because token will not be transferred from Portal to CDM. 

I don't really understand why we have a problem if we use another token?

Reply:
I'm sorry for my late reply.

My assumption is below.
- The user call relayMessage directly.
- Portal -> CDM approve
- The function will be fail (because if msg.sender is not Portal, **it is only permitted to replay failedMessage**)

Thank you always 🙇🏼‍♂️😍

Reply:
Oh, I'm sorry. All state will be reverted,,,
I'm sorry for my missing. I think this code is okay. But I want to think more about reentrancy attack. Let me reply as soon as again! 🙇🏼‍♂️

Reply:
I apologize for my strange thoughts earlier. Solidity is difficult for me,,,

But I think we can make this assumption.

1. User calls Portal's finalizeWithdrawal
- Call relayMessage from Portal
- Import TON with safeTransfer.
- relayMessage returns due to lack of gasFee (not to be reverted)
- Added to failedMessage.
2. User calls CDM's relayMessage
- The user re-executes the failed message.
- Return due to lack of gas money
- Repeatable

So I think we still need to move it inside the If statement as I said.

In case of replay message, it can only be executed by another account not Portal and the money must not be transferred.

Reply:
Excellent point, Steven
This scenario can happens.
Let me update it!

Reply:
Thank you very much! 🙇🏼‍♂️

Reply:
I updated!